pyteal/compiler/optimizer/optimizer.py

```diff
@@ -0,0 +1,89 @@
+from typing import cast, Set
+from pyteal.ir.tealblock import TealBlock
+
+from pyteal.ir.tealop import TealOp
+from pyteal.ir.ops import Op
+
+
+class OptimizeOptions:
+    """An object which specifies the optimizations to be performed and relevant context.
+    Args:
+
+        scratch_slots (optional): cancel contiguous store/load operations
+            that have no load dependencies elsewhere.
+        skip_ids (optional): the slot ids that should be skipped during
+            optimization. At the moment this includes:
+                1. reserved slots because they may have dependencies outside
+                the current application. For example, the 'gloads' opcode can
+                access the slots of other applications in the tx group.
+                2. global slots because they're outside the scope of global
+                optimizations, which only apply to the control flow graph of
+                a single subroutine.
+                3. slots used with dynamic scratch vars. These slots use
+                indirection by means of the 'stores' opcode and dependencies
+                can only be determined at runtime."""
+
+    def __init__(self, *, scratch_slots: bool = False, skip_ids: Set[int] = None):
```

Comment:
My understanding of `skip_ids` is that it's an implementation detail that's not meant to be user facing. If that's the case, I think it should not be possible to set in the constructor, and it should probably be prefixed with `_` so users know they shouldn't use it directly.

I'm suggesting this because I think in the long term, it _should_ be possible to pass information like this to optimizations in a different way, and if we make that change it would be nice if we could remove the field from here.

Reply:
Totally agree.